flight/tests/server.rs

```diff
@@ -0,0 +1,190 @@
+use std::{net::SocketAddr, str::FromStr as _, sync::Arc, time::Duration};
+
+use anyhow::{bail, Context as _, Result};
+use arrow::{compute::concat_batches, util::pretty::pretty_format_batches};
+use arrow_array::RecordBatch;
+use arrow_flight::{sql::client::FlightSqlServiceClient, FlightInfo};
+use arrow_schema::Schema;
+use ceramic_arrow_test::pretty_feed_from_batch;
+use ceramic_flight::{
+    server::{new_server, ConclusionFeed},
+    ConclusionData, ConclusionEvent, ConclusionInit, ConclusionTime,
+};
+use cid::Cid;
+use expect_test::expect;
+use futures::TryStreamExt as _;
+use http::Uri;
+use mockall::{mock, predicate};
+use tokio::net::TcpListener;
+use tonic::{async_trait, transport::Channel};
+
+/// Return a [`Channel`] connected to the TestServer
+#[allow(dead_code)]
+pub async fn channel(addr: &SocketAddr) -> Channel {
+    let url = format!("http://{addr}");
+    let uri: Uri = url.parse().expect("Valid URI");
+    Channel::builder(uri)
+        .timeout(Duration::from_secs(30))
+        .connect()
+        .await
+        .expect("error connecting to server")
+}
+
+async fn start_server(feed: MockFeed) -> FlightSqlServiceClient<Channel> {
+    let server = new_server(Arc::new(feed)).unwrap();
+    // let OS choose a free port
+    let listener = TcpListener::bind("127.0.0.1:0").await.unwrap();
+    let addr = listener.local_addr().unwrap();
+    let server_future =
+        server.serve_with_incoming(tokio_stream::wrappers::TcpListenerStream::new(listener));
+    tokio::spawn(server_future);
+    FlightSqlServiceClient::new(channel(&addr).await)
+}
+
+async fn execute_flight(
+    client: &mut FlightSqlServiceClient<Channel>,
+    info: FlightInfo,
+) -> Result<RecordBatch> {
+    let schema = Arc::new(Schema::try_from(info.clone()).context("valid schema")?);
+    let mut batches = Vec::with_capacity(info.endpoint.len() + 1);
+    batches.push(RecordBatch::new_empty(schema.clone()));
```

Comment:
why are we pushing an empty record batch here. Is it a placeholder for the schema? Can we add a comment here that explains the reasoning.

Reply:
This is test code, the empty batch ensures that when we print out the results the headers always print even if the results are empty. I'll add a comment